File: CSharp/5kyu/DirectionsReduction.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CSharp._5kyu
{
    public static partial class Kata5
    {
        public static string[] dirReduc(String[] arr)
        {
            var opposite = new Dictionary<string, string>{
                {"NORTH", "SOUTH"},
                {"EAST", "WEST"},
                {"SOUTH", "NORTH"},
                {"WEST", "EAST"},
            };

            var result = new List<string>();

            foreach (var dir in arr)
            {
                if (result.Count() > 0 && dir == opposite[result.LastOrDefault()])
                {
                    result.RemoveAt(result.Count() - 1);
                }
                else
                {
                    result.Add(dir);
                }
            }

            return result.ToArray<string>();
        }

        public static String[] dirReduc_kata_1(String[] arr)
        {
            Stack<String> stack = new Stack<String>();

            foreach (String direction in arr)
            {
                String lastElement = stack.Count > 0 ? stack.Peek().ToString() : null;

                switch (direction)
                {
                    case "NORTH": if ("SOUTH".Equals(lastElement)) { stack.Pop(); } else { stack.Push(direction); } break;
                    case "SOUTH": if ("NORTH".Equals(lastElement)) { stack.Pop(); } else { stack.Push(direction); } break;
                    case "EAST": if ("WEST".Equals(lastElement)) { stack.Pop(); } else { stack.Push(direction); } break;
                    case "WEST": if ("EAST".Equals(lastElement)) { stack.Pop(); } else { stack.Push(direction); } break;
                }
            }
            String[] result = stack.ToArray();
            Array.Reverse(result);

            return result;
        }
    }
}

File: JavaScript/5kyu/DirectionsReduction.js
function dirReduc(arr) {
  var directions = { NORTH: 0, WEST: 1, SOUTH: 0, EAST: 1 };

  var reducing = false;
  var start = true;
  while (reducing || start) {
    start = false;
    reducing = false;
    for (let i = 1; i < arr.length; i++) {
      if (
        arr[i - 1] != arr[i] &&
        directions[arr[i - 1]] == directions[arr[i]]
      ) {
        arr.splice(i - 1, 2);
        reducing = true;
        break;
      }
    }
  }

  return arr;
}

function dirReduc_kata_1(plan) {
  var opposite = {
    NORTH: "SOUTH",
    EAST: "WEST",
    SOUTH: "NORTH",
    WEST: "EAST",
  };

  return plan.reduce(function (dirs, dir) {
    if (dirs[dirs.length - 1] === opposite[dir]) dirs.pop();
    else dirs.push(dir);
    return dirs;
  }, []);
}

function dirReduc_kata_2(plan) {
  var str = plan.join("");
  var pattern = /NORTHSOUTH|EASTWEST|SOUTHNORTH|WESTEAST/;

  while (pattern.test(str)) {
    str = str.replace(pattern, "");
  }

  return str.match(/(NORTH|SOUTH|EAST|WEST)/g) || [];
}

var input = [
  "EAST",
  "EAST",
  "WEST",
  "NORTH",
  "WEST",
  "EAST",
  "EAST",
  "SOUTH",
  "NORTH",
  "WEST",
];
var result = dirReduc_kata_2(input);

console.log(result);
